internal/util/ssa/managedfields.go

```diff
@@ -20,86 +20,197 @@ package ssa
 import (
 	"context"
 	"encoding/json"
+	"time"
 
 	"github.com/pkg/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
 
+	clusterv1 "sigs.k8s.io/cluster-api/api/core/v1beta2"
 	"sigs.k8s.io/cluster-api/internal/contract"
 )
 
 const classicManager = "manager"
 
-// DropManagedFields modifies the managedFields entries on the object that belong to "manager" (Operation=Update)
-// to drop ownership of the given paths if there is no field yet that is managed by `ssaManager`.
-//
-// If we want to be able to drop fields that were previously owned by the "manager" we have to ensure that
-// fields are not co-owned by "manager" and `ssaManager`. Otherwise, when we drop the fields with SSA
-// (i.e. `ssaManager`) the fields would remain as they are still owned by "manager".
-// The following code will do a one-time update on the managed fields.
-// We won't do this on subsequent reconciles. This case will be identified by checking if `ssaManager` owns any fields.
-// Dropping ownership in paths for existing "manager" entries (which could also be from other controllers) is safe,
-// as we assume that if other controllers are still writing fields on the object they will just do it again and thus
-// gain ownership again.
-func DropManagedFields(ctx context.Context, c client.Client, obj client.Object, ssaManager string, paths []contract.Path) error {
-	// Return if `ssaManager` already owns any fields.
-	if hasFieldsManagedBy(obj, ssaManager) {
+// RemoveManagedFieldsForLabelsAndAnnotations removes labels and annotations from managedFields entries.
+// This func is e.g. used in the following context:
+// * First a full object is created with fieldManager.
+// * Then this func is called to drop ownership for labels and annotations
+// * And then in a subsequent syncMachines call a metadataFieldManager can take ownership for labels and annotations.
+func RemoveManagedFieldsForLabelsAndAnnotations(ctx context.Context, c client.Client, apiReader client.Reader, object client.Object, fieldManager string) error {
```

Comment:
q: should we add here
// Note: This is done in way that it does not rely on managedFields being stored in the cache, so we can optimize
// memory usage by dropping managedFields before storing objects in the cache.